coveo-functools/tests_functools/test_flex.py

```diff
@@ -0,0 +1,271 @@
+from dataclasses import dataclass
+from typing import Final, Dict, Any, Optional, Union, Callable, Generator, Type, Protocol
+
+import pytest
+from coveo_functools.exceptions import UnsupportedAnnotation
+from coveo_functools.flex import flex, RAW_KEY
+from coveo_testing.markers import UnitTest
+from coveo_testing.parametrize import parametrize
+
+
+@dataclass
+class MockInner:
+    value: str
+
+
+EXPECTED_VALUE = "SUCCESS"
+EXPECTED_VALUE_PAYLOAD = {"Value": EXPECTED_VALUE}
+PAYLOAD: Final[Dict[str, Any]] = {"Inner": EXPECTED_VALUE_PAYLOAD}
+EXPECTED_INNER: Final[MockInner] = MockInner(EXPECTED_VALUE)
+
+
+def _flex_mocks() -> Generator[Any, None, None]:
+    """General set of mocks meant to be called with DEFAULT_PAYLOAD."""
+
+    @flex
+    class MockClass:
+        def __init__(self, inner: MockInner) -> None:
+            self.inner = inner
+
+        @flex
+        def mock_method(self, inner: MockInner) -> MockInner:
+            return inner
+
+    yield MockClass
+    yield MockClass(EXPECTED_INNER).mock_method
+
+    @flex
+    @dataclass
+    class MockDataClass:
+        inner: MockInner
+
+    yield MockDataClass
+
+    @flex
+    def mock_function(inner: MockInner) -> MockInner:
+        return inner
+
+    yield mock_function
+
+
+@UnitTest
+@parametrize("obj", _flex_mocks())
+def test_flex_raw_data(obj: Any) -> None:
+    result = obj(**PAYLOAD)
+    assert getattr(result, RAW_KEY) == PAYLOAD

```

Comment:
Forget about this comment. From what I saw later, it's an advanced API feature :smile:

Reply:
Yes! When you get to a debugger view, any nested object would appear as something like `<module.Owner object at 0x00000000>` and that's not helpful at all. This is what happens if you inspect e.g. the "PyGithub" objects, you have to drill down everywhere to see what it is about, it's not convenient.

That's why I always put the raw payload on the object if I can.

At first it was a setting that you would enable (e.g.: `@flex(raw_key="my_raw_key")`) but I had a lot of troubles with the overloads/typing at the time and decided to not provide any configuration at all since mypy was way too confused. I had `strip_extras` as a setting too, dropped it. I might bring them back now that I learned a few more things about decorators vs class decorators.